Fix building without USE_DSTRING

tautschnig opened the pull request:
While there might be questions as to whether disabling USE_DSTRING is actually a good idea, all the ensuing fixes result in strictly cleaner code.

tautschnig committed:
Include unordered_map where using a std::unordered_map

The code fixed here previously relied on unordered_map being included by
dstring.h, which is only dragged in when USE_DSTRING is set.

tautschnig committed:
Do not use dstringt-specific APIs with irep_idt

Always use id2string instead of as_string, and directly refer to the string
container rather than indirectly via dstringt's API.

tautschnig committed:
Include string when not using dstringt as irep_idt representation

Fixes: #1837
together with all previous commits in this series.

romainbrenguier commented on src/goto-instrument/accelerate/acceleration_utils.h:
```diff
@@ -30,6 +31,7 @@ Author: Matt Lewis
 #include "cone_of_influence.h"
 
 typedef std::unordered_map<exprt, exprt, irep_hash> expr_mapt;
+typedef std::list<exprt> expr_listt;
```
This is used only once in this file so does not really need a `typedef`. Plus I think `std::list<exprt>` carries more information than `expr_listt`.

chrisr-diffblue replied:
Or alternatively, can the definition in `utils/expr.h` be used instead? It seems like `std::list<exprt>` is used all over the place in the cbmc repo, but the `expr_listt` typedef is already defined in that header file.

tautschnig replied:
I am wondering whether the typedef `expr_listt` provides much benefit? It's 6 characters shorter. Value would arise if anyone pondered swapping out `std::list` for some other implementation -- is that happening?

tautschnig replied:
For the specific instance here, however, I'd argue to keep the typedef: It's used 23 types in the accelerator code.

romainbrenguier commented on src/goto-symex/slice.h:
```diff
@@ -14,12 +14,16 @@ Author: Daniel Kroening, kroening@kroening.com
 // this simply slices away anything after the last assertion
 void simple_slice(symex_target_equationt &equation);
 
+typedef std::list<exprt> expr_listt;
```
same remark

romainbrenguier reviewed (approved):
Looks good

thk123 commented:
I haven't yet had a chance to properly review this - but so this doesn't break again, it would be really great to have a travis job compiling with dstring disabled.

tautschnig commented:
> I haven't yet had a chance to properly review this - but so this doesn't break again, it would be really great to have a travis job compiling with dstring disabled.

Would it be reasonable to merge this into "Ubuntu Linux with glibc using clang++-3.7, debug mode"?

thk123 commented:
> Would it be reasonable to merge this into "Ubuntu Linux with glibc using clang++-3.7, debug mode"?

This sounds good - certainly preferable I think to another job to wait for (and this seems primarily useful as  a debug tool anyway).

chrisr-diffblue commented:
> Would it be reasonable to merge this into "Ubuntu Linux with glibc using clang++-3.7, debug mode"?

Short of adding an extra configuration, I suspect that might be the best place to add it. Issue #1706 has some suggestions for changing around the travis configs, so might be also worth recording in that issue that we should consider USE_DSTRING as another variable in the matrix.

tautschnig commented:
> [...] so might be also worth recording in that issue that we should USE_DSTRING as another part of the matrix.

Done.

tautschnig committed:
Make USE_DSTRING conditional and disable it in Ubuntu/Clang/DEBUG Travis job

This introduces a macro USE_STD_STRING that can be set to disable USE_DSTRING,
and thus the use of dstringt as the implementation of irep_idt. The
Ubuntu/Clang/DEBUG Travis stage sets this macro to avoid future changes breaking
the build when disabling USE_DSTRING.

martin-cs commented:
I'm having a hard time seeing which of these commits actually meet the PR title and which are refactorings.  Fixing the build without USE_DSTRING ... I guess ...  I'd probably be OK with just removing the option.  Removing expr_listt feels like it should be a separate PR.  I'm less convinced by the value of that (again, small, wide changes without immediate links to a PR or feature).

tautschnig commented:
@martin-cs You are right, I'll split things off into separate PRs.

tautschnig committed:
Include list in expr.h

The code fixed here previously relied on list being included by dstring.h, which
is only dragged in when USE_DSTRING is set.

tautschnig commented:
@martin-cs I've removed unnecessary commits.

martin-cs reviewed (approved):
I guess approve but I do wonder what value this option adds.

tautschnig commented:
> [...] I do wonder what value this option adds.

I don't know - #1852.

thk123 commented:
One really simple advantage is `std::string` is supported out of the box by debuggers, whereas `dstringt` requires writing some kind of printer (and for GDB at least, the pretty printer we have seems to wander into some bug and cause the debugger to become unstable).

chrisr-diffblue commented on src/util/irep_ids.h:
```diff
@@ -16,6 +16,8 @@ Author: Reuben Thomas, reuben.thomas@me.com
 
```
To avoid the rather ugly perl in the travis config change that is part of this PR, is it worth considering tweaking the `#define USE_DSTRING` here to do something like:
```
#ifndef NO_USE_DSTRING
#define USE_DSTRING
#endif
```

Then the travis builds could just add `-DNO_USE_DSTRING` to the CXX_FLAGS?

I suppose also that if this change is made, it might be worth a quick update to `doc/architectural/cbmc-guide.md` as well, since that mentions `USE_DSTRING` briefly...

tautschnig replied:
I have now added `USE_STD_STRING` to guard `USE_STRING` in the way you suggested, and thus also got rid of the ugly perl hack.

chrisr-diffblue commented on .travis.yml:
```diff
@@ -166,6 +166,7 @@ jobs:
         - COMPILER="ccache clang++-3.7 -Qunused-arguments -fcolor-diagnostics"
         - CCACHE_CPP2=yes
         - EXTRA_CXXFLAGS="-DDEBUG"
+        - perl -p -i -e 's/^#define USE_DSTRING$//' src/util/irep_ids.h
```
Since this perl is rather ugly, is it worth tweaking the `#define USE_DSTRING` (see my other comment on src/util/irep_ids.h) to avoid needing to do this perl hack?

chrisr-diffblue reviewed (approved):
Happy with this now, though I note there is a CI failure which looks odd to me (a big pile of core dumps, by the look of it) which I suspect might be unrelated to this PR, so I'm approving on the safe assumption that you'll be sorting out the CI failure before merging :-)

tautschnig commented:
I have restarted the particular Travis stage.
